test/common/TestCommon.hs

```diff
@@ -0,0 +1,172 @@
+
+quickCheckTest0 :: TestName
+       -> (forall d r m. ( DualMonad d r m
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       -> ((Double, Double, Double) -> ([Double], [Double], [Double], [Double]))
+       -> TestTree
+quickCheckTest0 txt f fArg =
+  qcTestRanges txt f (listsToParameters4 . fArg) ((-2, -2, -2), (2, 2, 2)) ((-1e-7, -1e-7, -1e-7), (1e-7, 1e-7, 1e-7)) (-10, 10)
+
+-- A quick check to compare the derivatives and values of 2 given functions.
+cmpTwo
+  :: ( m ~ DualMonadForward r, d ~ 'DModeDerivative, Dual d r ~ r
+     , DualMonad d r m )
+  => (DualNumberVariables d r -> m (DualNumber d r))
+  -> (DualNumberVariables d r -> m (DualNumber d r))
+  -> Domains r
+  -> Domains r
+  -> Domains r
+  -> Domains r
+  -> Property
+cmpTwo f1 f2 params1 params2 ds1 ds2 =
+  close2 (dFastForward f1 params1 ds1) (dFastForward f2 params2 ds2)
+
+-- A quick check to compare the derivatives and values of 2 given functions.
+cmpTwoSimple
+  :: ( m ~ DualMonadForward r, d ~ 'DModeDerivative, Dual d r ~ r
+     , DualMonad d r m )
+  => (DualNumberVariables d r -> m (DualNumber d r))
+  -> (DualNumberVariables d r -> m (DualNumber d r))
+  -> Domains r
+  -> Domains r
+  -> Property
+cmpTwoSimple f1 f2 parameters ds =
+  cmpTwo f1 f2 parameters parameters ds ds
+
+-- A quick consistency check of all the kinds of derivatives and gradients
+-- and all kinds of computing the value of the objective function.
+qcPropDom :: (forall d r m. ( DualMonad d r m
+                            , r ~ Double
+                            , Floating (Out (DualNumber d (Vector r)))
+                            , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+              => DualNumberVariables d r -> m (DualNumber d r))
+       -> Domains Double
+       -> Domains Double
+       -> Domains Double
+       -> Double
+       -> Property
+qcPropDom f args ds perturbation dt =
+      let ff@(derivative, ffValue) = dFastForward f args ds
+          (derivativeAtPerturbation, valueAtPerturbation) = dFastForward f args perturbation
+          (gradient, revValue) = dReverse dt f args
+      in -- Two forward derivative implementations agree fully:
+         dForward f args ds === ff
+         -- Objective function value from gradients is the same.
+         .&&. ffValue == revValue
+         -- Gradients and derivatives agree.
+         .&&. close1 (dt * derivative)
+                     (dotParameters gradient ds)
+         -- Objective function value is unaffected by perturbation.
+         .&&. ffValue == valueAtPerturbation
+         -- Derivative approximates the perturbation of value.
+         .&&. close1 (primalValue
+                                  f (addParameters
+                                                   args perturbation))
```

Comment:
This is a completely borked indentation from the `sed` refactorings I was doing. Please don't bother fixing. I will do at some point.